impeller/toolkit/android/hardware_buffer.h

```diff
@@ -0,0 +1,127 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_IMPELLER_TOOLKIT_ANDROID_HARDWARE_BUFFER_H_
+#define FLUTTER_IMPELLER_TOOLKIT_ANDROID_HARDWARE_BUFFER_H_
+
+#include "flutter/fml/unique_object.h"
+#include "impeller/geometry/size.h"
+#include "impeller/toolkit/android/proc_table.h"
+
+namespace impeller::android {
+
+enum class HardwareBufferFormat {
+  //----------------------------------------------------------------------------
+  /// This format is guaranteed to be supported on all versions of Android. This
+  /// format can also be converted to an Impeller and Vulkan format.
+  ///
+  /// @see        Vulkan Format: VK_FORMAT_R8G8B8A8_UNORM
+  /// @see        OpenGL ES Format: GL_RGBA8
+  ///
+  /// Why have many format when one format do trick?
+  ///
+  kR8G8B8A8UNormInt,
+};
+
+using HardwareBufferUsage = uint8_t;
+
+enum class HardwareBufferUsageFlags : HardwareBufferUsage {
+  kFrameBufferAttachment = 1u << 0u,
+  kCompositorOverlay = 1u << 1u,
+  kSampledImage = 1u << 2u,
+};
+
+//------------------------------------------------------------------------------
+/// @brief      A descriptor use to specify hardware buffer allocations.
+///
+struct HardwareBufferDescriptor {
+  HardwareBufferFormat format = HardwareBufferFormat::kR8G8B8A8UNormInt;
+  ISize size;
+  HardwareBufferUsage usage = 0u;
+
+  //----------------------------------------------------------------------------
+  /// @brief      Create a descriptor of the given size that is suitable for use
+  ///             as a swapchain image.
+  ///
+  /// @warning    Descriptors of zero size are not allocatable. The next best
+  ///             valid size is picked. So make sure to check the actual size of
+  ///             the descriptor after this call is made to determine the size
+  ///             of the allocated hardware buffer.
+  ///
+  /// @param[in]  size  The size.
```

Comment:
Note size requirements here.

Reply:
I just put the size restrictions in the `@warning` block right above. Hoping to draw attention to it. But that was obviously not enough. I'll add a note reiterating the warning.